### PR TITLE
Simple LSTM wrapper with better default initialization

### PR DESCRIPTION
**Pull request by OyvindTafjord**
Not urgent, but this is an attempt at a mechanism for providing better default initialization for, say, LSTMs (and also suppress training of the superfluous `bias_hh` variables). I considered adding extra arguments to this new class so that it's easier to change the details, but that would be inconsistent with the other paradigm of parameter regexes global to the model. Maybe later though we'll also have module-specific initialization settings more generally, along the lines of Keras etc.

### CODE REVIEW

File: allennlp/modules/default_lstm.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    A wrapper of the Pytorch LSTM which provides a good default initialization
+    of weights and biases, and disables training of the superfluous bias_hh parameters.
+
+    Input parameters are identical to Pytorch LSTM.
+    """
+    @overrides
+    def reset_parameters(self):
+        hidden_size = self.hidden_size
+        for param_name, param in self.named_parameters():
+            if re.search("bias_hh", param_name):
```
Comment (matt-gardner) [inline allennlp/modules/default_lstm.py]: You shouldn't have to do a search here; you should get the name exactly, and can just match on that (`if name == "biah_hh":`).  Because you're calling `self.named_parameters()`, there's no nesting of names, or anything - you'll just get `"bias_hh"`, etc., as the names.

Comment (matt-gardner) [reply allennlp/modules/default_lstm.py]: In fact, you can just remove the whole loop over `self.named_parameters()`, and just directly initialize the things you want to initialize.  It's a tiny bit less robust to future changes in pytorch, but I think that's a feature, not a bug - if pytorch changes their LSTMs enough to break this, we will want to do things differently, too.

Comment (OyvindTafjord) [reply allennlp/modules/default_lstm.py]: There are the different layers and directions though, e.g., here's a full list for a 2-layer bidirectional one: `['weight_hh_l1_reverse', 'bias_hh_l1', 'bias_ih_l0', 'weight_ih_l1', 'weight_hh_l0', 'bias_ih_l1_reverse', 'bias_ih_l0_reverse', 'bias_hh_l1_reverse', 'weight_ih_l0', 'bias_hh_l0', 'weight_ih_l1_reverse', 'bias_ih_l1', 'weight_ih_l0_reverse', 'weight_hh_l0_reverse', 'weight_hh_l1', 'bias_hh_l0_reverse']`. I can generate that list, assuming nothing changes (and as you say, if it does, it might be a feature to find out), but this seemed cleaner.

**Comment by OyvindTafjord** [comment]
Yes, not thrilled with the `DefaultLstm` name either, and thought of just using `Lstm` or `InitializedLstm`. It depends a bit on what tasks we think such wrappers might take on (e.g., in Keras there are various regularizers/constraints/activations at this level as well).

File: allennlp/modules/default_lstm.py (hunk 1 above)
Comment (matt-gardner) [reply allennlp/modules/default_lstm.py]: Oh, yep, sorry, my bad.  You're right.

**Comment by matt-gardner** [comment]
I like `InitializedLstm` the most out of the options that I've seen.

File: allennlp/modules/default_lstm.py (hunk 1 above)
Comment (joelgrus) [reply allennlp/modules/default_lstm.py]: nit: it looks like in every case you're really just checking `param_name.startswith()`? if so, I feel like it's more readable that way.

File: allennlp/modules/default_lstm.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                torch.nn.init.constant(param, 0)
+                # This parameter is superfluous, so we should not update it in training
+                param.requires_grad = False
+            elif re.search("bias_ih", param_name):
+                bias = param.data
+                bias_size = bias.size(0)
+                # The forget get bias is between fractions 1/4 and 1/2 of the vector
+                start, end = bias_size//4, bias_size//2
+                bias.fill_(0)
+                bias[start:end].fill_(1)
```
Comment (DeNeutoy) [inline allennlp/modules/default_lstm.py]: can't it just be: `bias[hidden_size: 2 * hidden_size].fill_(1)`?

File: allennlp/modules/default_lstm.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    Input parameters are identical to Pytorch LSTM.
+    """
+    @overrides
+    def reset_parameters(self):
+        hidden_size = self.hidden_size
+        for param_name, param in self.named_parameters():
+            if re.search("bias_hh", param_name):
+                torch.nn.init.constant(param, 0)
+                # This parameter is superfluous, so we should not update it in training
+                param.requires_grad = False
```
Comment (DeNeutoy) [inline allennlp/modules/default_lstm.py]: I don't think this is worth the confusion for an initial user of having to do the `parameters = [p for p in self.model.parameters() if p.requires_grad]` trick were they to try to use this LSTM with their own training code. It's also pretty harmless.

File: allennlp/modules/default_lstm.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            elif re.search("bias_ih", param_name):
+                bias = param.data
+                bias_size = bias.size(0)
+                # The forget get bias is between fractions 1/4 and 1/2 of the vector
```
Comment (DeNeutoy) [inline allennlp/modules/default_lstm.py]: "forget gate"

File: allennlp/modules/default_lstm.py (hunk 1 above)
Comment (OyvindTafjord) [reply allennlp/modules/default_lstm.py]: Sure, I'll change that

File: allennlp/modules/default_lstm.py (hunk 2 above)
Comment (OyvindTafjord) [reply allennlp/modules/default_lstm.py]: Yes, that's clearer

File: allennlp/modules/default_lstm.py (hunk 3 above)
Comment (OyvindTafjord) [reply allennlp/modules/default_lstm.py]: Right, that might be confusing, I don't have a good sense of that. I'm not sure how harmless or not it is. It seems unnecessary to me to have hundreds of completely flat directions in your parameter space, although it's a small enough fraction of the total parameters that it practically might not matter much. Maybe I can make it a boolean flag which is off by default to avoid this annoyance with parameters (note we do this `requires_grad` filtering by default in our training code since we also have things like word embeddings which maybe be fixed, so it's probably general good practice to do this). We could also add a method to our `Model` class to get trainable parameters which does this filtering for you.

**Comment by OyvindTafjord** [comment]
OK, will change to `InitializedLstm` and make the other small suggested changes. I'm interested in opinions on whether we should, by default, disable gradients from the superfluous bias parameters or not.

**Comment by matt-gardner** [comment]
I vote for disabling it by default, with a warning in the class docstring, so no one stumbles into the error.  If they're using our code to get the `InitializedLstm` `from_params`, then they are probably using our training code too, which does the right thing.  If they're not using our framework, they will (hopefully) have read the docstring for `InitializedLstm`.

**Comment by OyvindTafjord** [comment]
I have made the discussed changes, but I propose just closing this PR for now and rather consider this pattern down the line if we decide more generally to add more custom initializations. Reasons:
   * Performance on BiDAF+SQuAD actually dropped (to EM 67.3 from EM 68.7 or so) with this change
   * For consistency it should be done more systematically across other modules, like GRU etc
   * We may decide on a different approach to initialization configuration
